### PR TITLE
plat-k3: drivers: Open TRNG firewall for TIFS

### PR DESCRIPTION
**Pull request by ti-kamlesh**
On devices with PLATFORM=k3-am62x, there is only one SA2UL instance, which is being shared between TIFS and OP-TEE.

Blocking access to TRNG from all other entities other than OP-TEE is causing firewall exception when being accessed by TIFS.

Allow access to TIFS to use SA2UL TRNG along with OP-TEE.

While at it, correct values for FW_SECURE_ONLY and FW_NON_SECURE.

### CODE REVIEW

File: core/arch/arm/plat-k3/drivers/sa2ul.c [hunk 1]
```diff
@@ -28,9 +28,10 @@
 #define FW_ENABLE_REGION        0x0a
 #define FW_BACKGROUND_REGION    BIT(8)
 #define FW_BIG_ARM_PRIVID       0x01
+#define FW_TIFS_PRIVID          0xca
 #define FW_WILDCARD_PRIVID      0xc3
-#define FW_SECURE_ONLY          GENMASK_32(8, 0)
-#define FW_NON_SECURE           GENMASK_32(16, 0)
+#define FW_SECURE_ONLY          GENMASK_32(7, 0)
+#define FW_NON_SECURE           GENMASK_32(15, 0)
```
Comment (manorit2001) [inline core/arch/arm/plat-k3/drivers/sa2ul.c:34]: Hi Kamlesh, I see that although both the fixes are related but are still not apt to come in a same commit. Probably split this fix out into a separate commit.

File: core/arch/arm/plat-k3/drivers/sa2ul.c [hunk 2]
```diff
@@ -116,10 +117,15 @@ static TEE_Result sa2ul_init(void)
 
 	/* Modify TRNG firewall to block all others access */
 	control = FW_ENABLE_REGION;
-	permissions[0] = (FW_BIG_ARM_PRIVID << 16) | FW_SECURE_ONLY;
 	start_address = RNG_BASE;
 	end_address = RNG_BASE + RNG_REG_SIZE - 1;
+	permissions[0] = (FW_BIG_ARM_PRIVID << 16) | FW_SECURE_ONLY;
+#if defined(PLATFORM_FLAVOR_am62x)
+	permissions[1] = (FW_TIFS_PRIVID << 16) | FW_NON_SECURE;
+	ret = ti_sci_set_fwl_region(fwl_id, rng_region, 2,
```
Comment (manorit2001) [inline core/arch/arm/plat-k3/drivers/sa2ul.c:125]: A stylistic comment but we should not split the function call like this when we can create a variable `num_permissions` and that might look cleaner.

**Review by manorit2001 — CHANGES_REQUESTED** [review]

File: core/arch/arm/plat-k3/drivers/sa2ul.c (hunk 2 above)
Comment (jenswi-linaro) [reply core/arch/arm/plat-k3/drivers/sa2ul.c:125]: +1

**Comment by glneo** [comment]
> On devices with PLATFORM=k3-am62x, there is only one SA2UL instance, which is being shared between TIFS and OP-TEE.

This is true for several TI platforms. You should be specific here on why you want this only for AM62.

**Comment by glneo** [comment]
> Hello, out of curiosity, how do you manage concurrent accesses to you TRNG entropy? Does your SA2UL module handles it so it can always give entropy to the requester?

No, it is still single user at a time. The only time TIFS uses the TRNG is during the low power mode exit sequence, and at that time OP-TEE is not active, the rest of the time it is exclusively used by OP-TEE. TIFS is our security controller, and already has control of all firewalls so it can be (has to be) trusted to do the right thing and not use its TRNG access to subvert OP-TEE's security.

**Review by manorit2001 — APPROVED** [review]
`Reviewed-by: Manorit Chawdhry <m-chawdhry@ti.com>`

**Comment by manorit2001** [comment]
Hi @jforissier , would this change be in 4.0.0 tag?

**Comment by jforissier** [comment]
> Hi @jforissier , would this change be in 4.0.0 tag?

Yes it will :)
